runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MetricResultsProtos.java

```diff
@@ -0,0 +1,235 @@
+        new ConcurrentHashMap<>();
+    private final Map<MetricKey, MetricResultBuilder<GaugeResult>> gauges =
+        new ConcurrentHashMap<>();
+
+    /**
+     * Populate metric-type-specific maps with {@link MonitoringInfo}s from a {@link
+     * BeamFnApi.MetricResults}.
+     */
+    public MetricResultsBuilder(BeamFnApi.MetricResults metrics) {
+      add(metrics.getAttemptedList(), false);
+      add(metrics.getCommittedList(), true);
+    }
+
+    private void add(Iterable<MonitoringInfo> monitoringInfos, Boolean committed) {
+      for (MonitoringInfo monitoringInfo : monitoringInfos) {
+        add(monitoringInfo, committed);
+      }
+    }
+
+    public MetricResults build() {
+      return new DefaultMetricResults(build(counters), build(distributions), build(gauges));
+    }
+
+    /** Helper for turning a map of result-builders into a sequence of results. */
+    private static <T> List<MetricResult<T>> build(Map<MetricKey, MetricResultBuilder<T>> map) {
+      return map.values().stream().map(MetricResultBuilder::build).collect(toList());
+    }
+
+    private void add(MonitoringInfo monitoringInfo, Boolean committed) {
+      add(
+          keyFromMonitoringInfo(monitoringInfo),
+          monitoringInfo.getType(),
+          monitoringInfo.getTimestamp(),
+          monitoringInfo.getMetric(),
+          committed);
+    }
+
+    private void add(
+        MetricKey metricKey, String type, Timestamp timestamp, Metric metric, Boolean committed) {
+      processMetric(
+          metric,
+          type,
+          timestamp,
+          counter -> add(metricKey, counter, committed, counters),
+          distribution -> add(metricKey, distribution, committed, distributions),
+          gauge -> add(metricKey, gauge, committed, gauges));
+    }
+
+    private <T> void add(
+        MetricKey key, T value, Boolean committed, Map<MetricKey, MetricResultBuilder<T>> map) {
+      if (committed) {
+        MetricResultBuilder builder = map.get(key);
+        if (builder == null) {
+          throw new IllegalStateException(
+              String.format("No attempted value found for committed metric %s: %s", key, value));
+        }
+        builder.setCommitted(value);
+      } else {
+        map.put(key, new MetricResultBuilder(key, value));
+      }
+    }
+
+    /**
+     * Helper class for storing and combining {@link MetricResult#getAttempted attempted} and {@link
+     * MetricResult#getCommitted committed} values, to obtain complete {@link MetricResult}s.
+     *
+     * <p>Used while traversing {@link MonitoringInfo} lists where they are stored separately.
+     */
+    private static class MetricResultBuilder<T> {
+      private final MetricKey key;
+      private final T attempted;
+      @Nullable private T committed;
+
+      public MetricResultBuilder(MetricKey key, T attempted) {
+        this.key = key;
+        this.attempted = attempted;
```

Comment:
Add a setAttempted for both
I would say the default is that committed are the only ones emitted by an SDK, which come from the ProcessBundleResponse, while attempted come from the ProcessBundleProgressResponse, requiring a lot more work.

Reply:
This is only used when we know we have an "attempted", and we may or may not have a "committed". All runners today populate "attempted", but only some populate "committed".

Encoding [the knowledge that we always have "attempted" here] makes `build` not have to do any checking; `attempted` is `final` and non-`@Nullable` (unlike `committed`, but that's ok because `MetricResult` also encodes this fact).

I found it cleaner than an approach that pretended to be more general (this is a `private static class` with just the logic it uses), though I understand calling it a `Builder` may be confusing / unidiomatic.

lmk if I should rename it, or still make it a more general "builder"-pattern, or what you think.

Reply:
There are a couple meanings of {"attempted", "committed"} floating around. 

You've implied they correspond to MIs that come in via {`ProcessBundleProgressResponse`, `ProcessBundleResponse`}, resp., but there's a pre-fn-api {"physical", "logical"} meaning that most existing code uses.

I'm not trying to resolve these two meanings in this PR. I wrote things to conform to exactly what was already happening, while still supporting the changes I'm attempting to make (which require converting between [new proto] and [existing SDK] representations of metrics).

What was already happening, that I'm preserving, is:
- "attempted" ("physical") is always set
- "committed" ("logical") is sometimes set
  - some runners don't support the latter, so they don't set it (by going through [`MCSM.asAttemptedOnlyMetricResults`](https://github.com/apache/beam/blob/3040767e0312fa73cd08d8f592ce958c484d9fc1/runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MetricsContainerStepMap.java#L114)):
    - [flink](https://github.com/apache/beam/blob/3040767e0312fa73cd08d8f592ce958c484d9fc1/runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainer.java#L146)
    - spark [1](https://github.com/apache/beam/blob/3040767e0312fa73cd08d8f592ce958c484d9fc1/runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/SparkBeamMetric.java#L47) [2](https://github.com/apache/beam/blob/3040767e0312fa73cd08d8f592ce958c484d9fc1/runners/spark/src/main/java/org/apache/beam/runners/spark/SparkPipelineResult.java#L109)
    - samza [1](https://github.com/apache/beam/blob/3040767e0312fa73cd08d8f592ce958c484d9fc1/runners/samza/src/main/java/org/apache/beam/runners/samza/metrics/SamzaMetricsContainer.java#L62) [2](https://github.com/apache/beam/blob/3040767e0312fa73cd08d8f592ce958c484d9fc1/runners/samza/src/main/java/org/apache/beam/runners/samza/SamzaPipelineResult.java#L86)
  - other runners do set both:
    - DataflowMetrics [1](https://github.com/apache/beam/blob/3040767e0312fa73cd08d8f592ce958c484d9fc1/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowMetrics.java#L143), [2](https://github.com/apache/beam/blob/3040767e0312fa73cd08d8f592ce958c484d9fc1/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowMetrics.java#L152): "attempted" and "committed" are set to the same value
    - [DirectMetrics](https://github.com/apache/beam/blob/3040767e0312fa73cd08d8f592ce958c484d9fc1/runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectMetrics.java#L259-L262): both values set, representing "physical" and "logical" values
    - [portable DirectMetrics](https://github.com/apache/beam/blob/3040767e0312fa73cd08d8f592ce958c484d9fc1/runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/DirectMetrics.java#L263-L266)
      - most portable DirectRunner metrics code seems like copy-pasta from the legacy version. I've consolidated / deleted some of it, but haven't gotten to this yet.
- accordingly, ["committed" is `@Nullable`](https://github.com/apache/beam/blob/3040767e0312fa73cd08d8f592ce958c484d9fc1/sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/MetricResult.java#L65-L66), and ["attempted" is not](https://github.com/apache/beam/blob/3040767e0312fa73cd08d8f592ce958c484d9fc1/sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/MetricResult.java#L68-L69). 

In this `private static` helper, I've written the simplest conversion I can manage between [an attempted metric value] and [optional corresponding committed value that comes in via a different proto field] and a MetricResult that joins them.

lmk if that doesn't make sense!